src/acquisition/flusurv/flusurv_update.py

```diff
@@ -68,122 +68,112 @@
-  )
-  ON DUPLICATE KEY UPDATE
-    `release_date` = least(`release_date`, %s),
-    `rate_age_0` = coalesce(%s, `rate_age_0`),
-    `rate_age_1` = coalesce(%s, `rate_age_1`),
-    `rate_age_2` = coalesce(%s, `rate_age_2`),
-    `rate_age_3` = coalesce(%s, `rate_age_3`),
-    `rate_age_4` = coalesce(%s, `rate_age_4`),
-    `rate_overall` = coalesce(%s, `rate_overall`),
-    `rate_age_5` = coalesce(%s, `rate_age_5`),
-    `rate_age_6` = coalesce(%s, `rate_age_6`),
-    `rate_age_7` = coalesce(%s, `rate_age_7`)
-  '''
-
-  # insert/update each row of data (one per epiweek)
-  for epiweek in epiweeks:
-    lag = delta_epiweeks(epiweek, issue)
-    if lag > 52:
-      # Ignore values older than one year, as (1) they are assumed not to
-      # change, and (2) it would adversely affect database performance if all
-      # values (including duplicates) were stored on each run.
-      continue
-    args_meta = [release_date, issue, epiweek, location, lag]
-    args_insert = data[epiweek]
-    args_update = [release_date] + data[epiweek]
-    cur.execute(sql, tuple(args_meta + args_insert + args_update))
-
-  # commit and disconnect
-  rows2 = get_rows(cur)
-  print('rows after: %d (+%d)' % (rows2, rows2 - rows1))
-  cur.close()
-  if test_mode:
-    print('test mode: not committing database changes')
-  else:
-    cnx.commit()
-  cnx.close()
+    """Fetch and store the currently avialble weekly FluSurv dataset."""
+
+    # fetch data
+    location_code = flusurv.location_codes[location_name]
+    print('fetching data for', location_name, location_code)
+    data = flusurv.get_data(location_code)
+
+    # metadata
+    epiweeks = sorted(data.keys())
+    location = location_name
+    release_date = str(EpiDate.today())
+
+    with Session(engine) as session:
+        rows1 = get_rows(session)
+        print('rows before: %d' % rows1)
+        for epiweek in epiweeks:
+            lag = delta_epiweeks(epiweek, issue)
+            if lag > 52:
+                # Ignore values older than one year, as (1) they are assumed not to
+                # change, and (2) it would adversely affect database performance if all
+                # values (including duplicates) were stored on each run.
+                continue
+            args_meta = {
+                "issue": issue,
+                "epiweek": epiweek,
+                "location": location,
+            }
+            args_update = {
+                "release_date": release_date,
+                "lag": lag,
+                "rate_age_0": data[epiweek][0],
+                "rate_age_1": data[epiweek][1],
+                "rate_age_2": data[epiweek][2],
+                "rate_age_3": data[epiweek][3],
+                "rate_age_4": data[epiweek][4],
+                "rate_overall": data[epiweek][5],
+                "rate_age_5": data[epiweek][6],
+                "rate_age_6": data[epiweek][7],
+                "rate_age_7": data[epiweek][8],
```

Comment:
question: is it worth refactoring `flusurv.extract_from_object` to produce a dict with `rate_*` keys instead of an array?

then including them here would be more straightforward

Reply:
I thought about it's complex as hell almost unreadabble  mostly cause of complex response from flusurv_url